### PR TITLE
Switch to web socket-based StreamingClientDAOs

### PR DESCRIPTION
**Pull request by mdittmer**
This is towards performance improvements for "cold boot" of web UI.

The plan is, roughly:

1. Use DAOs that can stream individual entities;
2. Use a worker to populate IndexedDB in the background
3. Update Angular components make more targeted queries (for "current page") against caching DAO

### CODE REVIEW

File: main/serve.js [hunk 1]
```diff
@@ -29,30 +31,51 @@ const ctx = foam.lookup('org.chromium.apis.web.DatastoreContainer').create({
...
 
-server.exportDirectory('/images', `${__dirname}/../static/images`);
-
-server.exportDirectory('/', `${__dirname}/../static/bundle`);
+// TODO(markdittmer): Explicitly select webSocketService port.
+// (Current default: 4000.)
+//
+// Start service by accessing lazily constructed property.
+ctx.webSocketService;
+console.log(` :: Serving web sockets from ${ctx.webSocketService.port}`);
```
Comment (mdittmer) [inline main/serve.js]: This should be `logger.log()`.

File: package.json [hunk 2]
```diff
@@ -26,7 +26,7 @@
     "angular": "^1.6.3",
     "angular-ui-router": "^0.4.2",
     "d3": "^4.7.4",
-    "foam2": "git://github.com/foam-framework/foam2.git#beta-1",
+    "foam2": "git://github.com/foam-framework/foam2.git#new-box",
```
Comment (mdittmer) [inline package.json]: Pending dependent FOAM2 PR. This change not to be checked in.

File: lib/client/api_service.es6.js [hunk 3]
```diff
@@ -23,33 +23,48 @@ angular.module('confluence').service('api', ['$window', function($window) {
...
-    of: org.chromium.apis.web.WebInterface,
+    of: WebInterface,
     daoType: 'MDAO',
   });
-  let releaseWebInterfaceJunctionDAO = foam.dao.RestDAO.create({
-    baseURL: RELEASE_API_URL,
-    of: org.chromium.apis.web.ReleaseWebInterfaceJunction,
-  });
+  let releaseWebInterfaceJunctionDAO = getClientDAO(
+      'releaseWebInterfaceJunctionDAO', ReleaseWebInterfaceJunction);
```
Comment (arobins) [inline lib/client/api_service.es6.js]: These strings should be constants defined in a shared file.

File: lib/controller/default.es6.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+'use strict';
+
+require('angular');
+
+// Provide controller for default view that will depend on API service.
+angular.module('confluence')
+  .controller('defaultController',
```
Comment (arobins) [inline lib/controller/default.es6.js]: nit: continuation indent is +4

File: lib/client/api_service.es6.js [hunk 5]
```diff
@@ -23,33 +23,48 @@ angular.module('confluence').service('api', ['$window', function($window) {
...
   const AGGRESSIVE_REMOVAL_URL = ORIGIN + '/aggressive-removal';
 
+  const ctx = foam.box.Context.create();
+  // Names must match those registered in in server box context.
+  function getClientDAO(name, cls) {
+    return foam.dao.StreamingClientDAO.create({
+      of: cls,
+      delegate: foam.box.SubBox.create({
+        name: name,
+        delegate: foam.box.WebSocketBox.create({
```
Comment (arobins) [inline lib/client/api_service.es6.js]: What's a Box? I looked at the source, and it's totally undocumented.

Comment (mdittmer) [inline lib/client/api_service.es6.js]: Hopefully this helps: https://github.com/foam-framework/foam2/blob/f66420803fc22bba315b64b7d24345ed17001893/src/foam/box/README.md

Comment (arobins) [inline lib/client/api_service.es6.js]: It does. Thanks.

File: lib/client/api_service.es6.js [hunk 6]
```diff
@@ -60,22 +70,21 @@ angular.module('confluence').service('api', ['$window', function($window) {
...
-  let browserSpecificDAO = foam.dao.RestDAO.create({
-    baseURL: BROWSER_SPECIFIC_URL,
-  });
-  let aggressiveRemovalDAO = foam.dao.RestDAO.create({
-    baseURL: AGGRESSIVE_REMOVAL_URL,
-  });
+  const ApiVelocityData = org.chromium.apis.web.ApiVelocityData;
+  const BrowserMetricData = org.chromium.apis.web.BrowserMetricData;
+  let apiVelocityDAO = getClientDAO(
+      container.API_VELOCITY_NAME,
```
Comment (arobins) [inline lib/client/api_service.es6.js]: nit: wrap this like the lines below

**Pull request merged**